torch/sparse/__init__.py

```diff
@@ -150,6 +150,48 @@
 """)
 
 
+spmm_reduce = _add_docstr(_sparse.spmm_reduce, r"""
+sparse.spmm_reduce(self, other, reduce, *, row_indices=None, ccol_indices=None, csr2csc=None) -> Tensor
+
+Reduce rows from dense matrices :attr:`other` at the locations specified by the sparsity pattern of :attr:`self`.
```

Comment:
Generally we try to preserve the invariant `op(<sparse_args>).to_dense() == op(<sparse_args>.to_dense())`. That is, the behavior of an operation is invariant under the layouts (strided, CSR, BSR, etc.) of the inputs.

Now that doesn't currently hold true for functions within the torch.sparse namespace (for example torch.sparse.sum's derivative is dependent on the indices of the sparse input argument).

Can this function be viewed as an extension of an already existing function? For example torch.sparse.mm or torch.sparse.hspmm?

Reply:
No, this one can not be viewed as extension of `torch.sparse.mm` or `torch.sparse.hspmm`. `spmm_reduce` supports a couple of reduce types such as `sum`, `mean`, `max` and `min`, this is not the same as torch.sparse.mm or torch.sparse.hspmm.

Reply:
@mingfeima To clarify my understanding, is it accurate to say that `reduce=sum` for `spmm_reduce` corresponds to a regular matmul between a sparse matrix $A$ and dense matrix $B$ where each element of the output is given by $output(i, k) = \sum_{j}A(i,j) * B(j,k)$,  and for instance `reduce=max` would correspond to $output(i, k) = \max_{j}A(i,j) * B(j,k)$? 

If this is the case, would it make sense to perhaps add a `reduce` flag to `torch.sparse.spmm`?

Reply:
Yes, mathematically they would be the same!

`torch.sparse.spmm` has been replaced by `torch.sparse.mm` now, so adding `reduce` in `torch.sparse.mm` ? In this way `torch.sparse.mm` would have different signature with `torch.mm`.

Also the backward performance would be a big issue, two reasons:
* `spmm_reduce` requires several meta info to be cached `row_indices`, `ccol_indices`, `csr2csc`. These tensors will be calculated from PyG side during data transformation in the DataLoader. IFF we need to re-calculate them in torch, it would be freakishly slow. Though we can directly transpose a SparseCsr tensor to SparseCsc in torch, it is quite slow right now because it relies on sorting the indices (and this `sort` is sequential on CPU and most GNN datasets would have extremely large size of indices).
* when the `reduce`  is amax or amin, we need to record the max/min value indices in the forward and pass them to the backward. So the autograd signature from `derivarives.ymal` of `spmm_reduce` is different from a standard `mm`. IF we re-calculate the indices in the backward, it would be roughly one magnitude slower. This is why I come up with an internal func called `_spmm_reduce(...) ->(Tensor, Tensor)` and wrap it up with `spmm_reduce(...) -> Tensor`.

So out of performance considerations, i would not like to mingle this one with `torch.mm`.

This one is more like a highway built for PyG, and end users would receive the benefit by using PyG (i do not expect that end users would directly use it from torch). it is originally named as [torch_sparse.spmm](https://github.com/rusty1s/pytorch_sparse/blob/master/torch_sparse/spmm.py), maybe we can come up with a new name that have more differentiation from `torch.mm` ? @rusty1s do you have any recommendations?

Reply:
@mingfeima - It's ok to extend torch.sparse.mm with this option. We also support sparse layouts for torch.mm, so we don't need the signature of torch.sparse.mm to match it. Furthermore, torch.sparse.mm's gradient is special so it by definition can't match torch.mm. 

On the point of `row_indices` and `ccol_indices`, we do have constructors for the CSR type that the user can call (which are zero copy) and then afterwards provide to `torch.sparse.mm`. Next, it's not true that `CSR.transpose` causes a call to sort. It should also be a zero-copy operation since we have the CSC layout.

Now, for `max/min` we can make `torch.sparse.mm` composite. For the `max/min` pass it'll call into `std::get<0>(_sparse_mm_tropical(<args>))` or such, which in turn has its own custom gradient in derivatives.yaml.

On your argument around "highway built for PyG", then why land it in PyTorch core? Now, if your goal is to reach a wide audience, then it's better to link into existing API surfaces. It's easier for the user to provide an input of a different layout than to rewrite the operators used in a potentially opaque model definition.

Reply:
also, i remember there was somewhere an old request for generalized matmul (even outside of sparse context): https://github.com/pytorch/pytorch/issues/59545 and my own hopes of having keops-like generalized matmul support in core

Reply:
how about we make an overload upon `torch.sparse.mm`, for example:

```
- func: _sparse_mm(Tensor sparse, Tensor dense) -> Tensor
  python_module: sparse
- func: _sparse_mm.reduce(Tensor sparse, Tensor dense, str reduce) -> Tensor
  python_module: sparse
```

Reply:
@cpuhrsch I have updated this patch to add an overload to `torch.sparse.mm` with an additional `reduce` argument, for example:

```
result = torch.sparse.mm(sparse, dense, 'max')
```

Also removed the doc of `spmm_reduce`.

Please review again! Thanks!